code/modules/mob/living/bot/bot.dm

```diff
@@ -441,5 +475,92 @@
+
+	for(var/obj/structure/railing/D in loc)
+		if(!D.density)			continue
+		if(dir & D.dir)			return 1
 	return 0
+
+
+//Multi-Z AStar Procs
+
+/turf/proc/Euclidean3dDistance(turf/t)
+	var/euclid_dist = sqrt(Square(src.x - t.x) + Square(src.y - t.y) + Square(src.z - t.z))
+	var/currentPathweight = src.pathweight
+	var/targetPathweight = t.pathweight
+	if(locate(/obj/machinery/door) in src)
+		currentPathweight += 1
+	if(locate(/obj/machinery/door) in t)
+		targetPathweight += 1
+	return euclid_dist * ((currentPathweight+targetPathweight)/2)
+
+/turf/proc/Manhattan3dDistance(turf/t)
+	var/manhattan_dist = abs(src.x - t.x) + abs(src.y - t.y) + abs(src.z - t.z)
+	var/currentPathweight = src.pathweight
+	var/targetPathweight = t.pathweight
+	if(locate(/obj/machinery/door) in src)
+		currentPathweight += 1
+	if(locate(/obj/machinery/door) in t)
+		targetPathweight += 1
+	return manhattan_dist * ((currentPathweight+targetPathweight)/2)
+
+//NORTH, SOUTH, EAST, WEST, NORTHEAST, NORTHWEST, SOUTHEAST, SOUTHWEST + Traversal up and down stairs
+/turf/proc/AllDirTurfsWithAccessWithZ(var/obj/item/weapon/card/id/ID)
+	var/L[] = new()
+
+	for(var/d in GLOB.alldirs)
+		var/turf/simulated/T = get_step(src, d)
+		if(istype(T) && !T.density)
+			var/turf/simulated/open/OP = T
+			if(istype(OP) && (!locate(/obj/structure/lattice) in OP) && !LinkBlockedWithAccess(src, OP, ID))
+				if(!OP.below.density)
+					L.Add(OP.below)
+			else if(!LinkBlockedWithAccess(src, T, ID))
+				L.Add(T)
+		else
+			var/obj/structure/stairs/S = locate(/obj/structure/stairs) in src
+			if(S?.dir == d && !LinkBlockedAbove(src,GetAbove(S),d))
+				L.Add(get_step(GetAbove(S), d))
+
+	return L
+
+//NORTH, SOUTH, EAST, WEST + Traversal up and down stairs
+/turf/proc/CardinalTurfsWithAccessWithZ(var/obj/item/weapon/card/id/ID)
+	var/L[] = new()
+
+	for(var/d in GLOB.cardinal)
+		var/turf/simulated/T = get_step(src, d)
+		if(istype(T) && !T.density)
+			var/turf/simulated/open/OP = T
+			if(istype(OP) && (!locate(/obj/structure/lattice) in OP) && !LinkBlockedWithAccess(src, OP, ID))
+				if(!OP.below.density)
+					L.Add(OP.below)
+			else if(!LinkBlockedWithAccess(src, T, ID))
+				L.Add(T)
+		else
+			var/obj/structure/stairs/S = locate(/obj/structure/stairs) in src
+			if(S?.dir == d && !LinkBlockedAbove(src,GetAbove(S),d))
+				L.Add(get_step(GetAbove(S), d))
+
+	return L
+
+//Checks for directional blockages at the base of the stairs leading up, open space above, and that nothing is blocking the exit point. Returns 1 on fail
+/proc/LinkBlockedAbove(var/turf/lower, var/turf/simulated/open/upper, var/dir)
+	if(!istype(upper))		return 1
+	for(var/obj/A in lower)
+		if(!A.density)		continue
+		if(istype(A,/obj/structure/window) || istype(A,/obj/structure/railing))
+			return (A.dir == dir)
+		else				return 1
+	var/turf/exit = get_step(upper, dir)
+	if(exit.density)		return 1
+	for(var/obj/B in exit)
+		if(B.density)		return 1
+	return 0
```

Comment:
I think this code is mostly unreachable. Either we return 1 in L561 because it's not a window or railing, or we return a boolean in L560 if it is. 

We'll only ever get to L562 if there are absolutely no objs in lower.

Reply:
Wow. No excuse for this, no idea how that return slipped past me! Corrected and added a reverse direction window check to upper rather than a flat density check.